src/rt/cover.c

```diff
@@ -368,323 +504,571 @@ static cover_file_t *cover_file(const loc_t *loc)
+         checked_sprintf(color, sizeof(color), "#00cc00");
+      } else if (perc > 90) {
+         checked_sprintf(color, sizeof(color), "#e6e600");
+      } else if (perc > 80) {
+         checked_sprintf(color, sizeof(color), "#ff9900");
+      } else {
+         checked_sprintf(color, sizeof(color), "#ff0000");
+      }
 
-      stats.total_stmts++;
+      fprintf(f, "      <td bgcolor=%s>%.1f %% (%d/%d)</td>\n",
+              color, perc, hit, total);
+      return;
    }
-   else {
-      const int start = tag->loc.first_column;
-      const int end = (tag->loc.line_delta == 0)
-         ? tag->loc.first_column + tag->loc.column_delta : l->len;
-
-      const int mask = (conds[tag->tag] >> (tag->sub_cond * 2)) & 3;
 
-      if (tag->sub_cond == 0) {
-         stats.total_branches++;
-         if (mask == 3)
-            stats.hit_branches++;
-      }
+   fprintf(f, "      <td bgcolor=#aaaaaa>N.A.</td>\n");
+}
 
-      stats.total_conds++;
-      if (mask == 3)
-         stats.hit_conds++;
+static void cover_print_hierarchy_header(FILE *f)
+{
+   fprintf(f, "<table style=\"width:70%%;margin-left:" MARGIN_LEFT "px;margin-right:auto;\"> \n"
+              "  <tr>\n"
+              "     <th bgcolor=#777777 style=\"width:40%%\">Instance</th>\n"
+              "     <th bgcolor=#777777 style=\"width:10%%\">Statement</th>\n"
+              "     <th bgcolor=#777777 style=\"width:10%%\">Branch</th>\n"
+              "     <th bgcolor=#777777 style=\"width:10%%\">Toggle</th>\n"
+              "     <th bgcolor=#777777 style=\"width:10%%\">Average</th>\n"
+              "  </tr>\n");
+}
 
-      cover_hl_t *hl;
-      for (hl = l->hl; hl != NULL; hl = hl->next) {
-         if ((hl->start == start) && (hl->end == end))
-            break;
-      }
+static void cover_print_hierarchy_footer(FILE *f)
+{
+   fprintf(f, "</table>\n");
+}
 
-      if (hl == NULL) {
-         hl = xmalloc(sizeof(cover_hl_t));
-         hl->start  = start;
-         hl->end    = end;
-         hl->next   = l->hl;
+static void cover_print_timestamp(FILE *f)
+{
+   time_t t;
+   time(&t);
 
-         l->hl = hl;
-      }
-      else if (hl->kind == HL_HIT)
-         return;
+   fprintf(f, "  </section>\n");
 
-      hl->kind = (mask == 3) ? HL_HIT : HL_MISS;
-      hl->help = NULL;
+   fprintf(f, "<footer>");
+   fprintf(f, "   <p> NVC version: %s </p>\n", PACKAGE_VERSION);
+   fprintf(f, "   <p> Generated on: %s </p>\n", ctime(&t));
+   fprintf(f,  "</footer>");
+}
 
-      if (mask == 1)
-         hl->help = "Condition never evaluated to TRUE";
-      else if (mask == 2)
-         hl->help = "Condition never evaluated to FALSE";
+static void cover_print_hierarchy_summary(FILE *f, cover_stats_t *stats, ident_t hier,
+                                          bool top)
+{
+   char dir[6] = {0};
+   if (top)
+      sprintf(dir, "hier/");
+
+   fprintf(f, "   <tr>\n"
+              "      <td><a href=\"%s%s.html\">%s</a></td>\n",
+              dir, istr(hier), istr(hier));
```

Comment:
Get rid of `dir` and just use `top ? "hier/" : ""`.

Reply:
Done